lib/attrsets.nix

```diff
@@ -534,23 +534,53 @@ rec {
         in mapAttrs g;
     in recurse [] set;
 
+  /* Generate an attribute set by mapping a function over a list of
+     attribute names.
+
+     Example:
+       mapListToAttrs (name: "x_" + name) [ "foo" "bar" ]
+       => { bar = "x_bar"; foo = "x_foo"; }
```

Comment:
`mapAttrsToList` should expectedly be the inverse of `mapListToAttrs`, but that's not the case here

Reply:
> Should `mapAttrsToList` only work on values instead?

`mapAttrsToList` is a ten year old function that works well. Changing it will do more harm than good.

A better framework for understanding the function is that it's `map f ∘ attrsToList`.

> With the preexisting built-in `mapAttrs` function working on only the values, we had to add a `mapAttrs'`.

> I personally believe either `builtins.mapAttrs` should effectively be `lib.mapAttrs'`,

The apostrophe generally refers to the function being either new, or more strict. Both apply in this case. `mapAttrs` can return the attribute set spine (`attrNames` + thunks) without evaluating any of the `attrValues`, whereas `mapAttrs'` is generally more strict, having to evaluate the function for each of the attributes. It does depend on the strictness of the passed function.
Defining a "map" function to operate on single values is quite conventional. It is originally a transformation of a function, although "multi-paradigm" languages may have you believe otherwise; in Nix it is a transformation of a function, and `mapAttrs` is more general, allowing any function to operate on attribute sets instead. `mapAttrs'` requires a curried function.

----------

We are well past the stage where we could have made such breaking changes to `lib`, let alone `builtins`. I think we're over ten years late for that.

A redesign would have to be done as `lib2` and would take a lot more than one change to make it worthwhile. Even then it has a huge cost for the ecosystem; an effort that would be better spent on something more productive.

The only reasonable thing we can do here is one or two functions that are consistent with the current patterns, which gets us back to the inverse of

    map f ∘ attrsToList

which would be

    listToAttrs ∘ map g

or equivalently

    mapAttrs h ∘ listToAttrs

In practice, the list items may not already be in the form of name-value pair attrsets, so the `listToAttrs ∘ map g` variant seems easier to use.
If the list items _are_ in a name-value pair form, then it's likely that another function such as `concatMapAttrs` could have done the job instead, so I'm inclined to dismiss that use case.

-------

For context, consistency in a library like this is very important, because remembering the signatures of individual functions is not feasible. Most of these functions do not pass the [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold) by themselves and are only a net positive because they form a consistent "language" of their own. This "language" only has somewhat of a right to exist for two reasons, which are
 - performance optimizations
 - lack of a truly convenient syntax for composing long chains of functions

Reply:
> The apostrophe generally refers to the function being either new, or more strict. Both apply in this case. mapAttrs can return the attribute set spine (attrNames + thunks) without evaluating any of the attrValues, whereas mapAttrs' is generally more strict, having to evaluate the function for each of the attributes. It does depend on the strictness of the passed function.

Okay, so by "strict", you mean in evaluation steps and in what it produces? I suppose that yes, it is more "general" in the sense that _all_ functions compliant with its type will work with attribute sets, though when I think "generality", I'm thinking of the data on which I'm operating, and what I can do with it, which I suppose would be "flexibility" instead, which mapAttrs' does have.

---

Similarly, I don't see how `listToAttrs ∘ map g` is equivalent to `mapAttrs h ∘ listToAttrs`. `listToAttrs` is predicated on the applied data set being pairs of name-value form, with `concatMapAttrs` operating on one existing attrset. I continually came across the issue of wanting to transform both name and value for lists, and kept running into not having an easy way of expressing it available.

I originally thought to take two separate functions and have it be of type `(A -> String) -> (A -> Any) -> [A] -> AttrSet`, but was pointed out it could be reduced, though IMO at the cost of immediate observable usability.

---

> For context, consistency in a library like this is very important...

Therein is my point. Consistency in _what_? Consistency in the strictness of evaluation? Consistency in the applicable data? Consistency in parameter ordering?

It's a tradeoff between simplicity and usability, no? You can have a library designed for the purpose of a strict language for package writing or testing, but that doesn't explain why the library functions available go beyond such. Isn't it for the developer to not only be able to shorten their code, but to be able to express their intent in different manners and conceptualize their mental models better?

Reply:
> Okay, so by "strict", you mean in evaluation steps and in what it produces?

Yes, strictness as in evaluation. A result is strict in an in argument when that argument must be evaluated in order produce the result.

-----

> Similarly, I don't see how `listToAttrs ∘ map g` is equivalent to `mapAttrs h ∘ listToAttrs`.

Equivalent in what the design permits to be expressed. The assumption here is that `g` and `h` are equivalent under some conversion, but I figured that was implied. I should've picked `f` and `f'` perhaps.

-----

> It's a tradeoff between simplicity and usability, no?

It's not; see the Fairbairn threshold link. In general, a single function is not easier to use than a composition of functions, unless the composition is non-trivial.

You may have noticed that any grouping of container-related functions can be refactored by abstracting the domain-specific code into parameters and giving the remainder a name. The Fairbairn threshold comes from roughly the observation that we'll eventually have as many instances of this glue code that the best naming scheme is to eventually just name all the functions after their implementations. From those implementations, only the functions with "significant" functionality need to remain and be part of the naming scheme. So instead of the user composing those significant functions, is it simpler or more usable to introduce a naming scheme that just repeats those functions? And that's assuming you get the naming scheme right and that's it's accurate and memorable. Actual code is already accurate and does not need to be memorable.

As I explained, in Nix we have other reasons to add such functions, which are a bit questionable from a usability and simplicity perspective; they're about keystrokes and - even more questionably - performance.

> Isn't it for the developer to not only be able to shorten their code

Indeed it's not. Shortening must not be the goal, but a positive side effect of other choices. Readability _is_ a goal, and shortening may or may not follow from that.

> but to be able to express their intent in different manners and conceptualize their mental models better?

For domain-specific code, yes, but this is not domain-specific code. `mapListToAttrs` conceptually does not abstract anything from "map + `listToAttrs`", except the order of the two, which is an arbitrary decision anyway. It being arbitrary indicates that it is not domain-specific and will lead to surprises because a user may expect the other decision. This is where consistency comes in.

-----

> Consistency in _what_?

Unwritten rules of thumb that explain why certain functions are named the way they are and how their inputs and outputs relate to similarly named functions. I wish I had a better answer, but that's how it is.

> Consistency in the strictness of evaluation?

As lazy as possible. This is only tangentially relevant here.

> Consistency in the applicable data?

If you'd optimize for this, you'd probably divide the library into sublibraries that operate largely on single types, and you'd reduce the functions that operate between the types to a minimum. This implies removing `mapAttrsToList`, but that ship has sailed.

> Consistency in parameter ordering?

Functions called "map" always take a function as their argument. ("first argument" if curried)

Reply:
I can see the reasoning in your replies @roberth, though I can't say I agree with the underlying assumption that developer usability and approachability is intertwined with experienced readability, or that there is one objectively sufficient one-size-fits-all implementation. I do see though that this is getting a little out-of-hand, and isn't as compatible with the existing nixpkgs library as I thought, so I might either simply add it to mine or factor out into my own library repository.

If this sounds amenable, I'll close the pull request or you can, at your discretion.